velox/expression/Expr.cpp

```diff
@@ -765,141 +708,61 @@ Expr::PeelEncodingsResult Expr::peelEncodings(
-        if (firstIndices) {
-          // having a mix of dictionary and sequence encoded fields
-          peeled = false;
-          break;
-        }
-        BufferPtr lengths = leaf->wrapInfo();
-        if (!firstLengths) {
-          firstLengths = std::move(lengths);
-        } else if (lengths != firstLengths) {
-          // different fields use different sequences
-          peeled = false;
-          break;
-        }
-        if (firstPeeled == -1) {
-          firstPeeled = fieldIndex;
-        }
-        setPeeled(leaf->valueVector(), fieldIndex, context, maybePeeled);
-      } else {
-        // Non-peelable encoding.
-        peeled = false;
-        break;
-      }
-    }
-    if (peeled) {
-      ++numLevels;
-      peeledVectors = std::move(maybePeeled);
+  std::vector<VectorPtr> vectorsToPeel;
+  vectorsToPeel.reserve(distinctFields_.size());
+  for (const auto& field : distinctFields_) {
+    auto fieldIndex = field->index(context);
+    assert(fieldIndex >= 0 && fieldIndex < numFields);
+    auto fieldVector = context.getField(fieldIndex);
+    if (fieldVector->isConstantEncoding()) {
+      // Make sure constant encoded fields are loaded
+      fieldVector = context.ensureFieldLoaded(fieldIndex, rows);
     }
-  } while (peeled && nonConstant);
-
-  if (numLevels == 0 && nonConstant) {
-    return Expr::PeelEncodingsResult::empty();
+    vectorsToPeel.push_back(fieldVector);
   }
 
-  // We peel off the wrappers and make a new selection.
-  SelectivityVector* newRows;
-  SelectivityVector* newFinalSelection;
-  if (firstPeeled == -1) {
-    // All the fields are constant across the rows of interest.
-    newRows = singleRow(newRowsHolder, rows.begin());
-    context.saveAndReset(saver, rows);
-    context.setConstantWrap(rows.begin());
-  } else {
-    auto decoded = localDecoded.get();
-    auto firstWrapper = context.getField(firstPeeled).get();
-    const auto& rowsToDecode =
-        context.isFinalSelection() ? rows : *context.finalSelection();
-    decoded->makeIndices(*firstWrapper, rowsToDecode, numLevels);
+  // Attempt peeling.
+  VELOX_CHECK(!vectorsToPeel.empty());
+  auto peeledEncoding = std::make_shared<PeeledEncoding>();
+  auto peeledVectors = peeledEncoding->getPeeledVectors(
+      vectorsToPeel, rowsToPeel, localDecoded, propagatesNulls_);
 
-    newRows = translateToInnerRows(rows, *decoded, newRowsHolder);
-
-    if (!context.isFinalSelection()) {
-      newFinalSelection = translateToInnerRows(
-          *context.finalSelection(), *decoded, finalRowsHolder);
-    }
-
-    context.saveAndReset(saver, rows);
+  if (isFlat(peeledEncoding->getWrapEncoding())) {
+    return Expr::PeelEncodingsResult::empty();
+  }
 
-    if (!context.isFinalSelection()) {
-      *context.mutableFinalSelection() = newFinalSelection;
-    }
+  // Translate the relevant rows.
+  SelectivityVector* newFinalSelection = nullptr;
+  if (!context.isFinalSelection()) {
+    VELOX_CHECK_NOT_NULL(context.finalSelection());
```

Comment:
nit: this should be not required right? or check should be inside context.isFinalSelection()